### DISCUSSION TITLE
Performance and Code size on MCU (esp32, 8266, etc.)

### DISCUSSION THREAD

**Discussion by jshep321** [nlohmann/json#2639] [Q&A]
Hi,
I'm trying to do some non-trival tree searching that I can't do with arduinojson, which works quite well but lacks a few features.

Has anyone tried this on an MCU such as the ESP32 (either native IDF or arduino framework)?  I'm curious of the heap performance and code size impacts.  

Thanks!

**Comment by anteph** [comment]
Hi! 

I've tested the library in the Arduino core (v2.0.0). I did not conduct any performance analysis as I'm mainly writing tutorials on the basics of using this library on the ESP32, but so far I did not encounter any issue during my tests.

Regarding code size, I don't have the numbers by heart, but I've used the nlohmann/json together with inja (a HTML template processor), the WiFi Stack and the HTTP server libraries, and did not encounter any issue as well in terms of exceeding the code size.

For reference, if someone eventually tries to use this lib as an Arduino library, at least with the current versions, I've faced a clash between a define in the Arduino core (binary.h) and some template definitions in the nlohmann/json single include file (the name in question was B1). A simple rename in the single header file did the trick and everything seems to be working nice :)

Thanks,
Nuno Santos

**Reply by nlohmann** [reply]
Thanks a lot for sharing (sorry I did not answer to your email earlier)! Is it just renaming all instances of `B1` or was there anything else to adjust? Furthermore, is there a way to add the used compiler to the CI? I never worked with an Arduino, but I would expect that there is some Linux-based toolchain that we could use.

**Comment by anteph** [comment]
Hi @nlohmann no worries, thanks for replying here :)

It was just adjusting all instances of B1 and making sure that the Arduino core for the ESP32 was on version 2.0.0. I put together a section on the blog post that I believe better captures everything I've done to have it running: https://techtutorialsx.com/2021/09/08/esp32-json/#Installing_the_Library.

Unfortunately I never setup any CI pipeline to automate stuff from Arduino / the ESP32 (I do mostly everything in the Arduino IDE), but from the Project page I believe they have implemented some automation: https://github.com/espressif/arduino-esp32.

The Arduino core for the ESP32 is built on top of IDF, the official development framework for the ESP32. There's some documentation about its toolchain: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/linux-setup.html. Not sure though if it would be easy to sort it out.

I can also confirm that I've tested the following features with success:
- Serialization
- Deserialization
- JSON flatten
- JSON diff / patch / merge patch

From the looks of it, I'm pretty optimistic that the rest of the features will work just fine :)

Thanks!
